### PR TITLE
add puzzle 15 example

### PR DESCRIPTION
**Pull request by GitVladlen**

### CODE REVIEW

File: examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    // utils
+    public:
+        ShapeQuadFixedPtr createSprite( const ConstString & _name, const ResourcePtr & _resource );
+        HotSpotPolygonPtr createHotSpot( const ConstString & _name, const mt::vec2f & _size );
+        NodePtr createGraphicsRect( const ConstString & _name, const Color & _color, const mt::vec2f & _size );
+        NodePtr createGraphicsRectFill( const ConstString & _name, const Color & _color, const mt::vec2f & _size );
+        TextFieldPtr createTextField( const ConstString & _name, const ConstString & _id, const VectorString & _args );
+
+        bool getImageSize( const FileGroupInterfacePtr & _fileGroup, const FilePath & _filePath, mt::vec2f * _out );
+        ResourceImageDefaultPtr createImageResource( const ConstString & _resourceName, const ConstString & _fileGroupName, const FilePath & _filePath, const mt::vec2f & _maxSize, mt::uv4f _uvImage, mt::uv4f _uvAlpha );
```
Comment (irov) [inline examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.h:58]: add cr

File: examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp [hunk 2]
```diff
@@ -0,0 +1,1097 @@
...
+    static MENGINE_CONSTEXPR const int32_t DIRECTION_DOWN = PUZZLE_COLS;
+    static MENGINE_CONSTEXPR const int32_t DIRECTION_LEFT = -1;
+    static MENGINE_CONSTEXPR const int32_t DIRECTION_RIGHT = 1;
+    //////////////////////////////////////////////////////////////////////////
+    static const FilePath & IMAGE_PATH()
+    {
+        return STRINGIZE_FILEPATH_LOCAL( "image_0.png" );
+    }
+    //////////////////////////////////////////////////////////////////////////
+    namespace Helper
```
Comment (irov) [inline examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp:52]: Detail?

File: examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp [hunk 3]
```diff
@@ -0,0 +1,1097 @@
...
+    {
+        //////////////////////////////////////////////////////////////////////////
+        MENGINE_INLINE void convertRowColToIdx( int32_t _row, int32_t _col, int32_t * _idx )
```
Comment (irov) [inline examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp:55]: static?

File: examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp [hunk 4]
```diff
@@ -0,0 +1,1097 @@
...
+
+        if( (_direction == DIRECTION_UP || _direction == DIRECTION_DOWN) && (fromCol != toCol) )
+        {
+            return;
+        }
+
+        Cook::addScope( _scope, this, &PuzzleSceneEventReceiver::scopeMakeTurn, toIdx );
+    }
+    //////////////////////////////////////////////////////////////////////////
+    bool PuzzleSceneEventReceiver::isGameOver()
```
Comment (irov) [inline examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp:988]: const

File: examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.h (hunk 1 above)
Comment (GitVladlen) [reply examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.h:58]: done

File: examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp (hunk 2 above)
Comment (GitVladlen) [reply examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp:52]: done

File: examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp (hunk 3 above)
Comment (GitVladlen) [reply examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp:55]: done

File: examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp (hunk 4 above)
Comment (GitVladlen) [reply examples/example_cpp_15_puzzle_demo/src/PuzzleFramework/PuzzleSceneEventReceiver.cpp:986]: done

**Pull request merged**